### PR TITLE
Project reporting to text

### LINKED ISSUE #18: Visualisation - Static report creation (text)

### PR DESCRIPTION
**Pull request by ThetaSinner**
resolves #18

### CODE REVIEW

File: acropolis-service/src/intTest/kotlin/org/ephyra/acropolis/service/ImportExportRoundTripIntTest.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.test.context.junit.jupiter.SpringExtension
+
+/**
+ * Test that the import and export services represent
```
Comment (Sub-Xaero) [inline acropolis-service/src/intTest/kotlin/org/ephyra/acropolis/service/ImportExportRoundTripIntTest.kt]: Represent what?

File: acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/SystemSoftwareRepository.kt [hunk 2]
```diff
@@ -16,4 +16,6 @@ internal interface SystemSoftwareRepository : CrudRepository<SystemSoftwareEntit
      * @return An instance of SystemSoftwareEntity if one is found
      */
     fun findByNameAndProjectId(name: String, projectId: Long): Optional<SystemSoftwareEntity>
+
+    fun findByProjectId(projectId: Long): List<SystemSoftwareEntity>
```
Comment (Sub-Xaero) [inline acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/SystemSoftwareRepository.kt]: Doc comment please

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IConnectionService.kt [hunk 3]
```diff
@@ -22,5 +22,5 @@ interface IConnectionService {
      * @param fromConnectable The instance of a Connectanle type to get all connections from
      * @return A list of all instances that have connections from the passed entity
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IConnectionService.kt]: `@param connectionType` comment please

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IApplicationSoftwareService.kt [hunk 4]
```diff
@@ -24,6 +24,13 @@ interface IApplicationSoftwareService {
      */
     fun find(name: String, projectId: Long): ApplicationSoftwareEntity?
 
+    /**
+     * Finds all the applications which belong to a project.
+     *
+     * @param projectName The name of the project to search in for applications
+     */
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IApplicationSoftwareService.kt]: `@return` comment? If we want to be thorough

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IExportService.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * Service interface for exporting project data
+ */
+interface IExportService {
+    /**
+     * Export a project identified by name to a specified data format.
+     *
+     * @param projectName The name of the project to export
+     * @param exportType The type of data format to output
+     */
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/IExportService.kt]: `@return` comment? If we want to be thorough

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/ISystemSoftwareService.kt [hunk 6]
```diff
@@ -23,6 +23,13 @@ interface ISystemSoftwareService {
      */
     fun find(name: String, projectId: Long): SystemSoftwareEntity?
 
+    /**
+     * Finds all the systems which belong to a project.
+     *
+     * @param projectName The name of the project to search in for systems
+     */
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/api/ISystemSoftwareService.kt]: `@return` comment? If we want to be thorough

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ApplicationSoftwareService.kt [hunk 7]
```diff
@@ -38,6 +38,17 @@ class ApplicationSoftwareService : IApplicationSoftwareService {
         return persistence.findByName(name, projectId)
     }
 
+    override fun findAll(projectName: String): List<ApplicationSoftwareEntity> {
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ApplicationSoftwareService.kt]: Do we want doc comments on the overrides?

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ExportService.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            ImportType.YAML -> {
+                yamlHelper.serialize(exportProject(project))
+            }
+            else -> {
+                throw IllegalStateException("Request to output to format [$exportType] which is not supported")
+            }
+        }
+    }
+
+    private fun exportProject(project: ProjectEntity): Project {
```
Comment (Sub-Xaero) [inline acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ExportService.kt]: Doc comments throughout.

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ApplicationSoftwareService.kt (hunk 7 above)
Comment (ThetaSinner) [reply acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ApplicationSoftwareService.kt]: Only if the override has details which are not covered by the interface documentation but need to be documented

File: acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ExportService.kt (hunk 8 above)
Comment (ThetaSinner) [reply acropolis-service/src/main/kotlin/org/ephyra/acropolis/service/impl/ExportService.kt]: Generally don't want to put doc comments on private methods. It's ignored by documentation tools so you would only do it informally for other developers to read while working on the project. If your private methods need doc comments then they're probably too complicated though!

File: acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/api/persistence/SystemSoftwarePersistence.kt [hunk 9]
```diff
@@ -44,9 +44,19 @@ class SystemSoftwarePersistence {
 
     /**
      * Updates an existing database record with the details contained within the entity parameter
-     * @param entity the instance of the entity containing modified fields to be persisted to the database
+     * @param systemSoftware The instance of the entity containing modified fields to be persisted to the database
```
[the remark this answers is not in the document]
Comment (ThetaSinner) [reply acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/api/persistence/SystemSoftwarePersistence.kt]: haha 👯‍♂️

File: acropolis-service/src/intTest/kotlin/org/ephyra/acropolis/service/ImportExportRoundTripIntTest.kt (hunk 1 above)
Comment (ThetaSinner) [reply acropolis-service/src/intTest/kotlin/org/ephyra/acropolis/service/ImportExportRoundTripIntTest.kt]: Is that not clear? :-S

**Comment by ThetaSinner** [comment]
@Sub-Xaero This I've got all the changes you suggested in :-) Thanks for the thorough review, I'd missed quite a few things!

**Pull request merged**